dartagnan/src/main/java/com/dat3m/dartagnan/encoding/NonTerminationEncoder.java

```diff
@@ -0,0 +1,760 @@
+        final Wmm memoryModel = task.getMemoryModel();
+        final RelationAnalysis ra = context.getAnalysisContext().requires(RelationAnalysis.class);
+        final BooleanFormulaManager bmgr = context.getBooleanFormulaManager();
+        final List<BooleanFormula> enc = new ArrayList<>();
+
+        // (1) Prefix-stores are co-before suffix stores
+        final Relation co = memoryModel.getRelation(CO);
+        final EventGraph coMay = ra.getKnowledge(co).getMaySet();
+        coMay.apply((x, y) -> {
+            if (isPossiblySuffix(y)) {
+                enc.add(bmgr.implication(
+                        bmgr.and(isInSuffix(y), isInPrefix(x)),
+                        bmgr.not(context.edge(co, y, x))
+                ));
+            }
+        });
+
+        final Relation rf = memoryModel.getRelation(RF);
+        final EventGraph rfMay = ra.getKnowledge(rf).getMaySet();
+        rfMay.apply((w, r) -> {
+            // (2) Nobody can read from suffix except suffix itself
+            if (isPossiblySuffix(w)) {
+                enc.add(bmgr.implication(
+                        bmgr.and(isInSuffix(w), context.edge(rf, w, r)),
+                        isInSuffix(r)
+                ));
+            }
+            // (3) Suffix can not read from prefix except from co-maximal prefix stores
+            if (isPossiblySuffix(r)) {
+                enc.add(bmgr.implication(
+                        bmgr.and(isInSuffix(r), context.edge(rf, w, r)),
+                        isSuffixReadable((Store)w)
+                ));
+            }
+        });
+
+        // (4) Events that can spuriously fail must succeed in the suffix (infinite spurious failure is unfair)
+        for (RMWStoreExclusive exclStore : task.getProgram().getThreadEvents(RMWStoreExclusive.class)) {
+            if (isPossiblySuffix(exclStore)) {
+                enc.add(bmgr.implication(isCfInSuffix(exclStore), context.execution(exclStore)));
+            }
+        }
+
+        return bmgr.and(enc);
+    }
+
+
+    // ------------------------------------------------------------------------------------------------------
+    // Helper variables for encoding
+
+    private BooleanFormula isInInfixVar(Iteration iter) {
+        return context.getBooleanFormulaManager().makeVariable(iter.getUniqueId() + "_infix");
+    }
+
+    private BooleanFormula isInSuffixVar(Iteration iter) {
+        return context.getBooleanFormulaManager().makeVariable(iter.getUniqueId() + "_suffix");
+    }
+
+    private BooleanFormula areInfixSuffixMatchingVar(Iteration a, Iteration b) {
+        final String varName = String.format("match(%s, %s)", a.getUniqueId(), b.getUniqueId());
+        return context.getBooleanFormulaManager().makeVariable(varName);
+    }
+
+    // ------------------------------------------------------------------------------------------------------
+    // Internal classes
+
+    /*
+        - A (possibly nonterminating) "Loop" contains a list of "NonterminationCase"s.
```

Comment:
How can more than one iteration be non-terminating? Or is it that you consider any iteration after the non terminating also non terminating?

Reply:
It's not more than one iteration, but there can be multiple ways in which a loop does not terminate. There are different iterations that can trigger a non-teminating situation and even multiple possiblities within a single iteration.
For example, if some loop/iteration is conditionally side-effect-free, then it may fail to terminate due to side-effect-free spinning (spinloop-tagged jump event) or due to side-effect-ful non-progress (other, nontermination-tagged jump).

Implementation-wise, the idea is that a single loop has special jump events tagged by `Tag.NONTERMINATION` to designate where non-termination could occur. Every such event is a "nontermination case" of the loop.